src/tasks/docker/package/package.js

```diff
@@ -157,11 +103,75 @@ const dockerPackage = async args => {
     }
   )
 
-  Logger.success(`Finished pushing docker image "${ imgTag }" to provider!`)
-  Logger.empty()
+}
 
-  return true
+/**
+ * Packages a docker container to be deployed to a docker providerCan 
+ * @function
+ * @param {Object} args - arguments passed from the runTask method
+ * @param {string} args.command - Initial command being run
+ * @param {Array} args.options - arguments passed from the command line
+ * @param {Object} args.tasks - All registered tasks of the CLI
+ * @param {Object} args.task - The current task being run
+ * @param {Object} args.params - Formatted options as an object
+ * @param {Object} globalConfig - Global config object for the keg-cli
+ *
+ * @returns {Object} - Build image as a json object
+ */
+const dockerPackage = async args => {
 
+  const { params, __internal=noOpObj } = args
+  const { action, author, context, message, push, } = params
+
+  const log = __internal.skipLog ? false : params.log
+
+  // Ensure we have a context to build the container
+  !context && throwRequired(args.task, 'context', args.task.options.context)
+
+  /* ---- Step 1: Get the container / image context info ---- */
+  const { contextEnvs, location, image, id } = await getResolvedContext(args)
+
+  /* ---- Step 2: Get the the image tags ---- */
+  const { imgWTag, cleanedTag, commitTag } = await getImgTags(params, location)
+
+  log && Logger.spacedMsg(`Creating docker package for ${imgWTag}...`)
+
+  /* ---- Step 3: Build the production app bundle for non-dev environments ---- */
+  action &&
+    await runBuildAction(id, args, {
+      action,
+      location,
+      container: id,
+      envs: contextEnvs,
+    }, true)
+
+  log && Logger.log(`Creating image from container ( ${id} )`)
+
+  /* ---- Step 4: Create image of the container using docker commit ---- */
+  const imageCreated = await imageFromContainer({
+    id,
+    log,
+    author,
+    imgWTag,
+    message,
+    cleanedTag,
+  })
+
+  /* ---- Step 5: Push docker image to docker provider registry ---- */
+  push &&
+    imageCreated &&
+    await pushImageToProvider(args, {
+      image,
+      imgWTag,
+      commitTag
+    }, log)
+
+  log && Logger.success(`Finished running docker package task!\n`)
+
+  // Return true so we know the image was pushed successfully
+  // We should only get here if all above steps are successful
+  // If they fail, they should throw
+  return true
```

Comment:
Currently we could remove this and things would be fine.
I do this because there are times when tasks are called by other tasks internally through the `runInternalTask` helper
In some cases the outcome of that task is then used by the calling task. This is used a lot within the `utils/services` files.

Probably the best example of this is `src/utils/services/startService.js`
* It calls the `pullService`, which calls the `docker pull task` and returns it's response
* The start service uses the response from the `pullService (i.e. docker pull task response)` to check if a new image was pulled.
  * It then uses that to determine if it should recreate  the docker containers if they are already running.
  * If there's no new image, then there's no reason to recreate the images if they are already running
    * So instead it just restarts them

Reply:
As for the `@throws`, I could add that, but I would say that's true for all tasks.
I'll update it, but I don't think it actually give you and extra information, because all tasks should throw if given invalid params / options

Reply:
Hmm, I think this is different. In the example you gave for the `pullService` function, the pull tasks are returning *objects* that `pullService` needs to return. It's not returning a boolean that `pullService` checks. 

The issue for me is not that a task can return values. The issue is the fact that this one is returning a boolean indicating that it succeeded. My point is that a caller of `dockerPackage` doesn't need it. They already know if the function succeeded or not, depending on whether or not it throws, and we should encourage them to use error handling.

I think having a function both throw and return a success-boolean can mislead a reader. They might think that since it returns a boolean, it might return false indicating a failure, when in fact it will never return false but instead throw.

Reply:
I have a number of thoughts on your comment, but I have to find the blog post / book I was reading to do it properly.
but there are reasons I throw in the tasks while also returning a true value.

In the case of the returning boolean of true, I just find it easier to reason about within code by using the return value to state the outcome.

In your example of `encourage them to use error handling`, nothing changes for the user. If they don't want the application to exit when calling the `package` task, they still have to catch the error, regardless of the `package` task returning true. So I don't see how this is any different.

If you don't care about the return value, then don't capture it, but either way, if the consumer has to catch the errors.
So you can still `encourage error handling`, you just also get `true` boolean if no error is thrown.

It ends up looking like this, without the boolean response =>
```js
runInternalTask(`docker.tasks.package`, {
  ...packageArgs
})
log(`We pushed a new package!`)
```
Vs with the boolean response =>
```js
const pushedNewPackage = runInternalTask(`docker.tasks.package`, {
  ...packageArgs
})

pushedNewPackage && log(`We pushed a new package!`)
```

For me personally,  If a dev looks only at this code
  * It's more clear that we only log the message if a new package was pushed
  * In the first example, one could think that we log the message regardless of the outcome of the `package` task

Granted in both cases the outcome would be the same, because if there's an error that task will thrown and exit. But if you don't know what the `package` task is doing, then I think the second option is more clear

All that being said, we are not actually using it, so I'm happy to remove the `return true` if you think it's needed

Reply:
So, I do agree that using return values is clearer in the calling context, because throwing is hidden unless the caller wraps it in a try catch. This is why I like `limbo` so much -- everything is very clear and the caller can still handle errors.

But I think that's an argument for just using return values, not using both. 

Because using both is slightly misleading. When I read this: 
```js
const pushedNewPackage = runInternalTask(`docker.tasks.package`, {
  ...packageArgs
})

pushedNewPackage && log(`We pushed a new package!`)

```
My first impression would be that the function does not throw. Because if it throws, there's no point in returning the boolean. I could easily imagine a developer modifying this to be:  
```js
const pushedNewPackage = runInternalTask(`docker.tasks.package`, {
  ...packageArgs
})

pushedNewPackage ? log(`We pushed a new package!`) : doSomethingElse()
```
which is dead code.

(Also, I do acknowledge throwing errors has major advantages too, but I'm still deciding how to make it work well in javascript. This blog was really interesting: https://overreacted.io/algebraic-effects-for-the-rest-of-us/).

Reply:
Your latest example makes a good point, but also misses the point of throwing, which is why I need to find that blog post. I just tried looking for it, but it was a while ago that I read it. I'll try searching this weekend.